### PR TITLE
correctly set test adapter when configure the queue adapter on a per job

### LINKED ISSUE #26360: [ActiveJob] test adapter is override by self.enque_adapter = :sidekiq
### Steps to reproduce

``` rb
class BidGoodsItemExpiredJob < ApplicationJob
  self.queue_adapter = :sidekiq
  def perform
       ....
  end
end
```

This will result in following test code not work.

``` rb
assert_enqueued_with(job: BidGoodsItemExpiredJob) do
      BidGoodsItemExpiredJob.set(wait_until: bid_goods_item.bid_expired_at).perform_later(bid_goods_item.id)
end
```

We must change job code to following to make test work.

``` rb
class BidGoodsItemExpiredJob < ApplicationJob
  self.queue_adapter = :sidekiq unless rails.env.test?
  def perform
       ....
  end
end
```

It weired, spent me hours to found out  test use 
`ActiveJob::QueueAdapters::TestAdapter` to make test work.
### Expected behavior

self.queue_adapter = :sidekiq in job should not change the default
test adapter ... it only test enque, not test enque to sidekiq.
### System configuration

Rails 5.0.0
Ruby 2.2.4

### PR DESCRIPTION
**Pull request by y-yagi**
### Summary

The `ActiveJob::TestHelper` replace the adapter to test adapter in `before_setup`.
It gets the target class using the `descendants`, but if the test target job
class is not loaded, will not be a replacement of the adapter.
To avoid this, modified to load the test target class.

Fixes #26360

### CODE REVIEW

File: activejob/lib/active_job/test_helper.rb [hunk 1]
```diff
@@ -10,6 +10,7 @@ module TestHelper
 
     def before_setup # :nodoc:
       test_adapter = queue_adapter_for_test
+      self.class.to_s.gsub(/Test\z/, "").safe_constantize
```
Comment (arthurnn) [inline activejob/lib/active_job/test_helper.rb]: This line is confusing, as we constantize something, and don't assign it to any variable.
Would be nice to add a comment in it, to make it more clear.

Comment (kaspth) [inline activejob/lib/active_job/test_helper.rb]: I'd write this as `self.class.name.chomp('Test').safe_constantize`.

What happens if users call `tests SomethingUnconventional`?
I found the original issue tough to grasp, but is this the best way we can find to solve the issue?

Comment (y-yagi) [inline activejob/lib/active_job/test_helper.rb]: > What happens if users call tests SomethingUnconventional?

`ActiveJob::TestCase` does not have `tests` method. So, calling `tests` will an error.  

> but is this the best way we can find to solve the issue?  

I think your concern is correct. I think again about the solution.

Comment (kaspth) [inline activejob/lib/active_job/test_helper.rb]: > `ActiveJob::TestCase` does not have `tests` method

People could include `ActiveJob::TestHelper` into an `ActiveSupport::TestCase` class and then it'd be there, iirc.

> I think your concern is correct. I think again about the solution.

❤️

**Comment by y-yagi** [comment]
I changed the solution.
When acquiring test adapter, it modified to return it if test adapter is set.

**Comment by arthurnn** [comment]
@kaspth thoughts?

**Review by rafaelfranca — APPROVED** [review]
@arthurnn :shipit:

**Comment by rafaelfranca** [comment]
Just avoid to use the conflict fix button on github because it creates an extra commit 😢

**Comment by arthurnn** [comment]
> Just avoid to use the conflict fix button on github because it creates an extra commit 😢

I rebased/squash in the end =)

**Pull request merged**